### PR TITLE
eclipses: Refactor eclipse plugin support to leverage p2

### PR DESCRIPTION
**Pull request by eliasvasylenko**
fixes #142191

What this PR does is rework how nixpkgs models a "plugin". (Note that there is a mismatch the language used by nixpkgs and Eclipse, specifically in what they refer to as a "plugin". But this already existed and can't really be fixed without creating confusing inconsistencies or just breaking backwards compat. And it's not really a big problem.)

Previously, the nixpkgs notion of a plugin was just a set of jars. These jars would be dumped into the dropins directory for an Eclipse instance.

The proposed nixpkgs notion of a plugin is a p2 update site, and an (optional) set of ids and versions which index into that update site. This gives Eclipse more information to help it resolve a working runtime.

This allows us to build a plugin derivation by mirroring a p2 update site locally. The problem with this is that a p2 update site is not a fixed resource, so the hash may change! An artifact in a p2 update site, for a given id and version, *is* specified to be a fixed resource, and so the hash shouldn't change, but artifacts can be added to an update site over time. The solution to this problem is to perform only a *partial* mirror, extracting a feature with a specific version and closing over its dependencies. I believe this should be reliable in practice, as the feature->plugin dependencies used are specified with exact versions.

###### Motivation for this change

Practical issues with the previous approach are outlined in #142191. More positively:

- This approach more closely aligns with the Eclipse model as its intended to be used, for what that's worth.
- I think it actually makes packaging eclipse plugins easier in most cases, as package maintainers can just use public-facing update sites as intended, rather than searching for zipped update sites (which aren't always available) or manually trawling for the transitive closure of jar dependencies which constitute a feature.
- The changes should be forwards compatible with 95% existing derivations out in the wild.

###### Problems with this change

There are probably some users out there with custom plugin derivations which will stop working, but hopefully it shouldn't be too much work to fix them. Specifically, derivations which directly call `buildEclipsePluginsBase` won't work, but I think that should be rare, as it's not a user-friendly mechanism to begin with.

There will be other edge-cases too. For instance jsonedit needed changing due to the dependency on the antlr runtime. Since there is no official packaging for the antlr runtime, the jsonedit project repackages it themselves and publishes it in the jsonedit p2 repo. This is fine, but they specify the dependency in such a way that you *have to* use their repackaged version to make the Eclipse resolver happy. This wasn't a problem with the old dropins model as the dropins mechanism is a bit looser with the rules for wiring everything up.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

Since there will be a small number of breakages downstream, does this PR warrant release notes? I have tried to keep it compatible as far as possible; most users of Eclipse shouldn't notice.

I've manually tested the changes by configuring an Eclipse instance with the following plugins:
- `bndtools`: to make sure #142191 is actually fixed.
- `jsonedit`, `antlr-runtime_4_5`, `antlr-runtime_4_7`: as these were the only existing plugins which needed changing.
- `yedit`: to check that `buildEclipsePlugin` still works.
- `checkstyle`: to check that `buildEclipseUpdateSite` still works.

They all appeared to be correctly installed and working.

I'm not an expert in Nix packaging so I expect some questionable choices will surface in review! I'll put some of my own comments inline where I wasn't sure about things.

### CODE REVIEW

File: pkgs/applications/editors/eclipse/default.nix [hunk 1]
```diff
@@ -29,6 +29,18 @@ in rec {
             makeWrapper;
   };
 
+  ### Eclipse Platform
```
Comment (eliasvasylenko) [inline pkgs/applications/editors/eclipse/default.nix:32]: On reflection this move was unnecessary, but it's benign. The motivation I suppose is that this is the "most minimal" eclipse, so it goes first! It may also be possible to refactor the others to build on top of this one using the same mechanisms that are used to install plugins, which would simply things a bit ... but this is probably an "if it ain't broke don't fix it" situation.

File: pkgs/applications/editors/eclipse/plugins.nix [hunk 2]
```diff
@@ -1,9 +1,28 @@
-{ lib, stdenv, fetchurl, fetchzip, unzip }:
+{ lib, stdenv, fetchurl, fetchzip, zip, unzip, xmlstarlet, eclipses }:
 
 rec {
 
-  # A primitive builder of Eclipse plugins. This function is intended
-  # to be used when building more advanced builders.
+  sanitizeUpdateSite = outDir: ''
```
Comment (eliasvasylenko) [inline pkgs/applications/editors/eclipse/plugins.nix:5]: This feels a bit ad hoc.

Both these jars just contain a single xml file, and the xml files contain an attribute with a timestamp, so we:
- extract
- zero the xml timestamp with xmlstarlet
- clear file timestamps
- rezip

Is there a cleaner way to do any of this? Existing utilities in nixpkgs to help?

File: pkgs/applications/editors/eclipse/plugins.nix [hunk 3]
```diff
@@ -15,80 +34,123 @@ rec {
 
       passthru = {
         isEclipsePlugin = true;
+        dependencies = [];
```
Comment (eliasvasylenko) [inline pkgs/applications/editors/eclipse/plugins.nix:37]: Dependencies are just a list of other plugins (which are of course local p2 repos). When we install our plugins these repos are just added to the set known to the resolver. Not sure if this is an appropriate use of passthru, or if there is a better way to express any of this.

File: pkgs/applications/editors/eclipse/plugins.nix [hunk 4]
```diff
@@ -15,80 +34,123 @@ rec {
...
+        then [ installableUnit ]
+        else installableUnits;
+      iuEntries = (lib.strings.concatStrings (map (iu: "<iu id=\"${iu.id}\" version=\"${iu.version}\" />") ius));
+
+    in
+      buildEclipsePluginBase rec {
+        name = "${pname}-${version}";
+        inherit version;
+
+        buildCommand = ''
```
Comment (eliasvasylenko) [inline pkgs/applications/editors/eclipse/plugins.nix:69]: There is a simpler way to mirror repositories than using an ant task... but unfortunately it mirrors the entire repository which means we can't have a reliable output hash. The ant task allows us to "slice" the repository, i.e. include only the specified installable units (features/plugins).

File: pkgs/applications/editors/eclipse/plugins.nix [hunk 5]
```diff
@@ -15,80 +34,123 @@ rec {
...
+        outputHashMode = "recursive";
+
+        passthru = {
+          installableUnits = map (iu: "${iu.id}/${iu.version}") ius;
+        };
+
+        inherit repository artifactRepository metadataRepository;
+      };
+
+  # Publish the given plugin and feature jars as a p2 update site.
```
Comment (eliasvasylenko) [inline pkgs/applications/editors/eclipse/plugins.nix:104]: This shouldn't be used under normal circumstances, but it needs to be here for two reasons.
- Backwards compatibility.
- Plugins which aren't published in p2 repos. This is *extremely* rare these days in the Eclipse ecosystem!

File: pkgs/applications/editors/eclipse/plugins.nix [hunk 6]
```diff
@@ -135,20 +197,14 @@ rec {
     };
   };
 
-  antlr-runtime_4_5 = buildEclipsePluginBase rec {
+  antlr-runtime_4_5 = buildEclipsePlugin rec {
```
Comment (eliasvasylenko) [inline pkgs/applications/editors/eclipse/plugins.nix:200]: Not sure why these were using buildEclipsePluginBase to begin with, buildEclipsePlugin would always have been simpler. But this is now a forced move, since buildEclipsePluginBase doesn't work any more.

File: pkgs/applications/editors/eclipse/plugins.nix [hunk 7]
```diff
@@ -223,6 +273,19 @@ rec {
     };
   };
 
+  bndtools = mirrorEclipsePlugins rec {
```
Comment (eliasvasylenko) [inline pkgs/applications/editors/eclipse/plugins.nix:276]: This is added as an example of using the new system as intended, and also as a sanity check since it was the original motivating example which didn't work with the previous system.

File: pkgs/applications/editors/eclipse/plugins.nix [hunk 8]
```diff
@@ -437,34 +500,16 @@ rec {
     };
   };
 
-  jsonedit = buildEclipsePlugin rec {
-    name = "jsonedit-${version}";
+  jsonedit = mirrorEclipsePlugins rec {
```
Comment (eliasvasylenko) [inline pkgs/applications/editors/eclipse/plugins.nix:503]: As mentioned in the root comment of the PR, this change was needed due to peculiarities with the antlr dependency. I think it makes a good case that the new system is easier to use though!

File: pkgs/applications/editors/eclipse/default.nix [hunk 9]
```diff
@@ -29,6 +29,18 @@ in rec {
...
+  eclipse-platform = buildEclipse {
+    name = "eclipse-platform-${platform_major}.${platform_minor}";
+    description = "Eclipse Platform ${year}-${month}";
+    src =
+      fetchurl {
+        url = "https://www.eclipse.org/downloads/download.php?r=1&nf=1&file=/eclipse/downloads/drops${platform_major}/R-${platform_major}.${platform_minor}-${timestamp}/eclipse-platform-${platform_major}.${platform_minor}-linux-gtk-x86_64.tar.gz";
+        sha512 = "29hab3ha3spk0048k3mf2x5m80hlh1l6nazsykx0xxrqy9vdkdibv6mq74dzf1n93h1bd5qh9axicnhs465kp7r1irdl04761c1wibi";
+      };
+  };
+
```
Comment (eliasvasylenko) [inline pkgs/applications/editors/eclipse/default.nix:43]: On reflection this move was unnecessary, but it's benign. The motivation I suppose is that this is the "most minimal" eclipse, so it goes first! It may also be possible to refactor the others to build on top of this one using the same mechanisms that are used to install plugins, which would simply things a bit ... but this is probably an "if it ain't broke don't fix it" situation.

File: pkgs/applications/editors/eclipse/plugins.nix [hunk 10]
```diff
@@ -1,9 +1,28 @@
...
+    xmlstarlet edit \
+      --inplace \
+      --update "/repository/properties/property[@name='p2.timestamp']/@value" \
+      --value "0" \
+      artifacts.xml content.xml
+    touch artifacts.xml content.xml -t '197001010100'
+    zip -X - artifacts.xml > "${outDir}/artifacts.jar"
+    zip -X - content.xml > "${outDir}/content.jar"
+    popd
+  '';
```
Comment (eliasvasylenko) [inline pkgs/applications/editors/eclipse/plugins.nix:20]: This feels a bit ad hoc.

Both these jars just contain a single xml file, and the xml files contain an attribute with a timestamp, so we:
- extract
- zero the xml timestamp with xmlstarlet
- clear file timestamps
- rezip

Is there a cleaner way to do any of this? Existing utilities in nixpkgs to help?

File: pkgs/applications/editors/eclipse/plugins.nix [hunk 11]
```diff
@@ -15,80 +34,123 @@ rec {
...
+                  <slicingOptions followStrict="true"/>
+                </p2.mirror>
+              </target>
+            </project>' > /build/build.xml
+          ${eclipse}/bin/eclipse \
+            -noSplash \
+            -configuration /build/configuration \
+            -application org.eclipse.ant.core.antRunner \
+            -buildfile /build/build.xml
+        '' + (sanitizeUpdateSite "$out");
```
Comment (eliasvasylenko) [inline pkgs/applications/editors/eclipse/plugins.nix:91]: There is a simpler way to mirror repositories than using an ant task... but unfortunately it mirrors the entire repository which means we can't have a reliable output hash. The ant task allows us to "slice" the repository, i.e. include only the specified installable units (features/plugins).

File: pkgs/applications/editors/eclipse/plugins.nix [hunk 12]
```diff
@@ -15,80 +34,123 @@ rec {
...
+        '';
 
-            mkdir -p $dropinDir/plugins
-            for plugin in ${toString pSrcs}; do
-              cp -v $plugin $dropinDir/plugins/$(stripHash $plugin)
-            done
-          '';
-        });
+        nativeBuildInputs = [ eclipse ];
+      });
```
Comment (eliasvasylenko) [inline pkgs/applications/editors/eclipse/plugins.nix:146]: This shouldn't be used under normal circumstances, but it needs to be here for two reasons.
- Backwards compatibility.
- Plugins which aren't published in p2 repos. This is *extremely* rare these days in the Eclipse ecosystem!

File: pkgs/applications/editors/eclipse/plugins.nix [hunk 13]
```diff
@@ -135,20 +197,14 @@ rec {
...
     };
 
-    buildCommand = ''
-      dropinDir="$out/eclipse/dropins/"
-      mkdir -p $dropinDir
-      cp -v $src $dropinDir/${name}.jar
-    '';
-
     meta = with lib; {
       description = "A powerful parser generator for processing structured text or binary files";
```
Comment (eliasvasylenko) [inline pkgs/applications/editors/eclipse/plugins.nix:209]: Not sure why these were using buildEclipsePluginBase to begin with, buildEclipsePlugin would always have been simpler. But this is now a forced move, since buildEclipsePluginBase doesn't work any more.

File: pkgs/applications/editors/eclipse/plugins.nix [hunk 14]
```diff
@@ -223,6 +273,19 @@ rec {
...
+
+    repository = "https://bndtools.jfrog.io/bndtools/update-latest";
+    installableUnits = [
+      { id = "bndtools.main.feature.feature.group"; version = version; }
+      { id = "bndtools.m2e.feature.feature.group"; version = version; }
+      { id = "bndtools.pde.feature.feature.group"; version = version; }
+    ];
+    hash = "sha256:163l714zgx2r5yn19m8jsc9jb9rkhq6215n0ml52fiymkkxcfc8k";
+  };
+
```
Comment (eliasvasylenko) [inline pkgs/applications/editors/eclipse/plugins.nix:288]: This is added as an example of using the new system as intended, and also as a sanity check since it was the original motivating example which didn't work with the previous system.

File: pkgs/applications/editors/eclipse/plugins.nix [hunk 15]
```diff
@@ -437,34 +500,16 @@ rec {
...
-          { n = "text"; h = "0clywylyidrxlqs0n816nhgjmk1c3xl7sn904ki4q050amfy0wb2"; }
-        ];
-
-    propagatedBuildInputs = [ antlr-runtime_4_7 ];
+    repository = "https://boothen.github.io/Json-Eclipse-Plugin";
+    installableUnits = [
+      { id = "jsonedit-feature.feature.group"; version = version; }
+      { id = "org.antlr.v4.feature.group"; version = "4.7.1"; }
+    ];
+    hash = "sha256:0af5d26lw1yxck5w73wkv57fpjx49pcx58gpflbrbxl3m5pc6js1";
```
Comment (eliasvasylenko) [inline pkgs/applications/editors/eclipse/plugins.nix:512]: As mentioned in the root comment of the PR, this change was needed due to peculiarities with the antlr dependency. I think it makes a good case that the new system is easier to use though!

File: pkgs/applications/editors/eclipse/default.nix (hunk 9 above)
Comment (jerith666) [reply pkgs/applications/editors/eclipse/default.nix:43]: Note that this introduces a (trivial) conflict with https://github.com/NixOS/nixpkgs/pull/150757; you just need the new hash in this new location.

**Comment by jerith666** [comment]
Do you intend your author name to be a bare `=` like this?

```
$ git log -1 6ca57d323bf909fe058bc141a6f95222360933ef | grep ^Author
Author: = <eliasv@saiman.co.uk>
```

**Comment by jerith666** [comment]
I have a few plugins locally that I hadn't pushed to nixpkgs; I just opened https://github.com/NixOS/nixpkgs/pull/153244 to get them in.

But, one of them doesn't work with this PR.  I think it's due to a whitespace issue.

```
$ nix-build -I nixpkgs=$(pwd) -E 'with import <nixpkgs> {}; with pkgs.eclipses; eclipseWithPlugins { eclipse = eclipse-platform; plugins = [ plugins.mat ]; }' -K
...
Unknown option Analyzer. Use -help for a list of known options.
There were errors. See log file: /build/configuration/1641151187969.log

$ grep Analyzer /tmp/nix-build-eclipse-platform-4.22.drv-0/configuration/1641151187969.log --color
Framework arguments:  -application org.eclipse.equinox.p2.director -destination /nix/store/6zia9d3k5gxn3agl1xkssfmv46lyaxcy-eclipse-platform-4.22/eclipse -repository file:/nix/store/vspm6lm42bs9scfbh9gky03mybgpd576-eclipse-plugin-mat-1.9.1 -installIU 201908260659.Memory Analyzer IDE,201908260659.Memory Analyzer RCP,org.eclipse.mat.api,org.eclipse.mat.chart,org.eclipse.mat.chart.feature.feature.group,org.eclipse.mat.chart.feature.feature.jar,org.eclipse.mat.chart.ui,org.eclipse.mat.dtfj,org.eclipse.mat.feature.feature.group,org.eclipse.mat.feature.feature.jar,org.eclipse.mat.hprof,org.eclipse.mat.ibmdumps,org.eclipse.mat.jdt,org.eclipse.mat.jruby.resolver,org.eclipse.mat.parser,org.eclipse.mat.report,org.eclipse.mat.ui,org.eclipse.mat.ui.help,org.eclipse.mat.ui.rcp,org.eclipse.mat.ui.rcp.feature.feature.group,org.eclipse.mat.ui.rcp.feature.feature.jar,org.eclipse.mat.ui.rcp.feature_root,org.eclipse.mat.ui.rcp.feature_root.gtk.linux.ppc64,org.eclipse.mat.ui.rcp.feature_root.gtk.linux.x86,org.eclipse.mat.ui.rcp.feature_root.gtk.linux.x86_64,org.eclipse.mat.ui.rcp.feature_root.win32.win32.x86,org.eclipse.mat.ui.rcp.feature_root.win32.win32.x86_64, -tag AddIUs
Command-line arguments:  -os linux -ws gtk -arch x86_64 -application org.eclipse.equinox.p2.director -destination /nix/store/6zia9d3k5gxn3agl1xkssfmv46lyaxcy-eclipse-platform-4.22/eclipse -repository file:/nix/store/vspm6lm42bs9scfbh9gky03mybgpd576-eclipse-plugin-mat-1.9.1 -installIU 201908260659.Memory Analyzer IDE,201908260659.Memory Analyzer RCP,org.eclipse.mat.api,org.eclipse.mat.chart,org.eclipse.mat.chart.feature.feature.group,org.eclipse.mat.chart.feature.feature.jar,org.eclipse.mat.chart.ui,org.eclipse.mat.dtfj,org.eclipse.mat.feature.feature.group,org.eclipse.mat.feature.feature.jar,org.eclipse.mat.hprof,org.eclipse.mat.ibmdumps,org.eclipse.mat.jdt,org.eclipse.mat.jruby.resolver,org.eclipse.mat.parser,org.eclipse.mat.report,org.eclipse.mat.ui,org.eclipse.mat.ui.help,org.eclipse.mat.ui.rcp,org.eclipse.mat.ui.rcp.feature.feature.group,org.eclipse.mat.ui.rcp.feature.feature.jar,org.eclipse.mat.ui.rcp.feature_root,org.eclipse.mat.ui.rcp.feature_root.gtk.linux.ppc64,org.eclipse.mat.ui.rcp.feature_root.gtk.linux.x86,org.eclipse.mat.ui.rcp.feature_root.gtk.linux.x86_64,org.eclipse.mat.ui.rcp.feature_root.win32.win32.x86,org.eclipse.mat.ui.rcp.feature_root.win32.win32.x86_64, -tag AddIUs
!MESSAGE Unknown option Analyzer. Use -help for a list of known options.
```

Note that the comma-separated list supplied to the `-installIU` option starts like:

```
201908260659.Memory Analyzer IDE,201908260659.Memory Analyzer RCP,org.eclipse.mat.api,org.eclipse.mat.chart
```

and I'm guessing the spaces in `Memory Analyzer IDE` and `Memory Analyzer RCP` are confusing it.

**Comment by jerith666** [comment]
```
--- a/pkgs/applications/editors/eclipse/default.nix
+++ b/pkgs/applications/editors/eclipse/default.nix
@@ -178,7 +178,7 @@ in rec {
           ${p2Director} \
             -destination $out/eclipse \
             -repository '${repository}' \
-            -installIU ${installIU} \
+            -installIU "${installIU}" \
             -tag AddIUs
           cat << EOF >> $out/eclipse/eclipse.ini
           ${jvmArgsText}
```

handles that, and lo and behold, I get a real error now:

```
Installation failed.
Cannot complete the install because one or more required items could not be found.
        Software being installed: Memory Analyzer for Eclipse IDE 1.0.0.47Y3cLT2nh7E7Ag7cEU7UQg (201908260659.Memory Analyzer IDE 1.0.0.47Y3cLT2nh7E7Ag7cEU7UQg)
        Missing requirement: Memory Analyzer (Charts) [optional] 1.9.1.201908260659 (org.eclipse.mat.chart.feature.feature.group 1.9.1.201908260659) requires 'org.eclipse.equinox.p2.iu; org.eclipse.birt.chart.engine 2.3.0' but it could not be found
        Cannot satisfy dependency:
                From: Memory Analyzer for Eclipse IDE 1.0.0.47Y3cLT2nh7E7Ag7cEU7UQg (201908260659.Memory Analyzer IDE 1.0.0.47Y3cLT2nh7E7Ag7cEU7UQg)
                To: org.eclipse.equinox.p2.iu; org.eclipse.mat.chart.feature.feature.group [1.9.1.201908260659,1.9.1.201908260659]
```

This looks like another example of the whole point of this exercise, yes?  If so, I like it!  :)

**Comment by jerith666** [comment]
Ah, and if I instead try to install the MAT plugin on top of `eclipse-jee`, it satisfies that dependency but then fails on another one:

```
$ nix-build -I nixpkgs=$(pwd) -E 'with import <nixpkgs> {}; with pkgs.eclipses; eclipseWithPlugins { eclipse = eclipse-jee; plugins = [ plugins.mat ]; }' -K
...
Installation failed.
Cannot complete the install because one or more required items could not be found.
        Software being installed: Memory Analyzer for Eclipse IDE 1.0.0.47Y3cLT2nh7E7Ag7cEU7UQg (201908260659.Memory Analyzer IDE 1.0.0.47Y3cLT2nh7E7Ag7cEU7UQg)
        Missing requirement: Memory Analyzer (Charts) [optional] 1.9.1.201908260659 (org.eclipse.mat.chart.feature.feature.group 1.9.1.201908260659) requires 'org.eclipse.equinox.p2.iu; org.eclipse.birt.chart.engine 2.3.0' but it could not be found
        Cannot satisfy dependency:
                From: Memory Analyzer for Eclipse IDE 1.0.0.47Y3cLT2nh7E7Ag7cEU7UQg (201908260659.Memory Analyzer IDE 1.0.0.47Y3cLT2nh7E7Ag7cEU7UQg)
                To: org.eclipse.equinox.p2.iu; org.eclipse.mat.chart.feature.feature.group [1.9.1.201908260659,1.9.1.201908260659]
```

But, it says it's an optional dependency ... so should it really be failing?

**Comment by rycee** [comment]
I wonder if this change perhaps needs some editing of [`eclipse.section.md`](https://github.com/NixOS/nixpkgs/blob/master/doc/builders/packages/eclipse.section.md)? Also, maybe it is of sufficient significance to add a brief note in the 22.05 release notes?

**Comment by jerith666** [comment]
Yeah, I think I agree that we need some documentation updates.  The errors I'm seeing are, I think, good progress.  But, it's not at all obvious to me how to go about resolving them.  I think I need to declare the plugin's dependencies explicitly somehow, but I don't know how to do that.  And I expect other users might run into similar problems.

**Comment by eliasvasylenko** [comment]
@rycee, yes you're right about those docs. Believe it or not I wasn't aware of them! Thank you for enlightening me, I'll push a commit with some appropriate changes. Same goes for the release notes.

@jerith666 Thank you for taking the time to look at this, and for catching the whitespace/quote mistake.

As for the issue installing the MAT plugin, I believe that's working as intended. (Don't worry the problem is solvable.)

There are a few key points here:

- The `[optional]` part of the `Memory Analyzer (Charts) [optional]` feature is just a part of the name string. Meaning that it's intended to communicate optionality to the user and not to the resolver.
- And in this case the user has elected to install it, and the resolver doesn't see this choice as optional. So if it can't install we get an error.

The behaviour of `buildEclipseUpdateSite` is to try to install *all* features and plugins at the update site, so naturally it tries to install the charts feature too. With the previous system it would have dumped the charts feature and all related plugin jars into the dropins dir and then just quietly failed to load them at runtime. The new system will tell you up-front that there is a problem.

So there are two options, we can either redefine the MAT plugin to only install only the main feature...
```
  mat = plugins.mirrorEclipsePlugins rec {
    pname = "mat";
    version = "1.9.1.201908260659";

    repository = "https://archive.eclipse.org/mat/1.9.1/update-site/";
    installableUnit = "org.eclipse.mat.feature.feature.group";
    hash = "sha256-usnlmTuLA/pAwh1706vwWtrqA4rPc4PGpODTq/aXWbw=";

    meta = with lib; {
      homepage = "https://www.eclipse.org/mat/";
      description = "The Eclipse Memory Analyzer is a fast and feature-rich Java heap analyzer that helps you find memory leaks and reduce memory consumption";
      license = licenses.epl10;
      platforms = platforms.all;
    };
  };
```

or we can give it enough info to find the correct dependencies...
```
  birt = plugins.mirrorEclipsePlugins rec {
    pname = "birt";
    version = "4.7.0.v201706222054";

    repository = "https://download.eclipse.org/releases/2021-12/";
    installableUnits = [
      "org.eclipse.birt.chart.device.extension"
      "org.eclipse.birt.chart.device.swt"
      "org.eclipse.birt.chart.engine.extension"
      "org.eclipse.birt.chart.engine"
      "org.eclipse.birt.core"
      { id = "org.mozilla.javascript"; version = "1.7.5.v201504281450"; }
    ];
    hash = "sha256-Gqcn0CjYMGbh2a4jPlEiEkr1zIWxWos7FW/Ac3QeP84=";
  };

  mat = plugins.mirrorEclipsePlugins rec {
    pname = "mat";
    version = "1.9.1.201908260659";

    repository = "https://archive.eclipse.org/mat/1.9.1/update-site/";
    installableUnits = [
      "org.eclipse.mat.feature.feature.group"
      "org.eclipse.mat.chart.feature.feature.group"
    ];
    hash = "sha256-3Xb61DruYJelI5GRQFCanaJnQu+Xzfw+5WfT+UPmch4=";

    dependencies = [ birt ];

    meta = with lib; {
      homepage = "https://www.eclipse.org/mat/";
      description = "The Eclipse Memory Analyzer is a fast and feature-rich Java heap analyzer that helps you find memory leaks and reduce memory consumption";
      license = licenses.epl10;
      platforms = platforms.all;
    };
  };
```

(Note that I had to push a few changes to get this to work.)

**Comment by eliasvasylenko** [comment]
So you do have to jump through some hoops to get MAT working with the chart feature. It's awkward that you manually have to wire up the dependency on Birt.

In normal use of Eclipse when you install a plugin they can resolve dependencies from any of the update sites configured on the eclipse instance. The release repo `https://download.eclipse.org/releases/2021-12/` would be configured by default, and so it'd be queried for the required BIRT plugins under the hood when you try to install MAT and everything would just work.

For now I think my way is fine, everything will work if you put a little extra energy in, you just have to be more explicit about dependencies. If you'll indulge me to speculate on some future work that could help here though...

We could achieve better ergonomics in this case if we could find a version of the release repo that had a fixed content hash and use that as a default dependency for plugins. I believe this would also have allowed you to install MAT directly on "eclipse-platform" and not just "eclipse-jee".

Now, the release repo is a composite repo, and I suspect the component repos, e.g. `https://download.eclipse.org/releases/2021-12/202112081000/` have fixed content, but I can't find any documentation which guarantees this. We could probably also find a snapshot archive of it somewhere.

An alternative would be to build the release repo from source, which would be pretty cool. In turn we could define the different eclipses, "eclipse-jee" etc, in a consistent way in terms of "epp" meta-packages which depend on the release repo.

One thing at a time though.

**Comment by jerith666** [comment]
Thanks, those examples are super helpful.  I'm trying to grok how you arrived at them ... I guess you have to install the plugin manually in order to figure out the value for the version (e.g. `4.7.0.v201706222054` for birt), right?  And it takes some iteration to find each missing dependency in turn as well, yes?

**Comment by eliasvasylenko** [comment]
The workflow is pretty bad for manually finding dependencies to be honest. Yes one approach is just installing something manually and searching the list of installed packages to see what ends up getting installed. Something else I find myself doing is manually scanning through feature and package listings under update sites. Unfortunately not all third-party update sites provide a useful web view, and I'm not aware of any other (convenient) tools for querying them.

One reason things are so awkward is that I've smooshed together the concepts of "update site" and "plugin set to install" into a single entity. It's kind of difficult to describe the limitations and possible tradeoffs succinctly, but basically there are two steps to the process of preparing and installing a set of plugins:
- Determine the subset of a p2 repository to mirror locally. This does *not* use the resolver, and doesn't properly chase dependencies into other repos, meaning more manual iteration is needed.
- Determine the plugins from that repository to install. This *does* use the resolver, but can only operate over the repositories and plugins we've actually mirrored.

As I alluded to in the last comment, it would be possible to have better ergonomics for resolving dependencies (generally the same experience as manually installing plugins in Eclipse), by supporting mirroring important repositories in their entirety instead of just subsets. The costs of choosing to mirror an entire repo this way would be:

- Much higher disk use during installation.

- Difficulty of finding snapshot of repositories with a stable hash.

But this would be on an opt-in basis. Simple plugins and those already defined could continue to work the same way, but those with dependencies on plugins in the release repo, like MAT, would be much easier to define if the release repo were available in its entirety.

Sorry for dumping all these huge walls of text on you one after another, it might be a bit of an overload. But talking it out has helped me work through some things, I think I'll take a crack at separating out the "repository" and "plugin" concepts, see if I can land on a model that's easier to work with.

**Comment by jerith666** [comment]
Thanks, that makes sense.  I agree it would be great to find a stable update site for each quarterly release and satisfy dependencies from there automatically somehow.

(And I for one appreciate a good wall of text -- I'm learning a lot by following this issue!  :) )

**Comment by jerith666** [comment]
I did find a couple of small clean-ups for your patch; I've pushed them here: https://github.com/eliasvasylenko/nixpkgs/compare/eclipse-plugins-using-p2...jerith666:eclipse-plugins-using-p2

**Comment by bjornfor** [comment]
@eliasvasylenko: Do you want to pick this up again? (Currently conflicts with master.)
